### ISSUE TITLE
Delayed Job downtime on Katuma due to subscriptions placement job

### ISSUE THREAD

**Issue by sauloperez** [openfoodfoundation/openfoodnetwork#7475]
## Description

Happy Apps reported downtime for Delayed Job on 04/20/2021 at 08:02 UTC which lasted until 08:13 AM UTC. Any job taking >5min to process is considered downtime.

As I was working at the time of the incident I could tail the logs and see it happen in real-time. Our single DJ worker was busy placing subscription orders, and each of them was quite slow to process. I could see the log hang for a bit after outputting each message. It turns out that those with issues take even longer to process.

<details>
<summary>Excerpt of log/delayed_job.log</summary>

openfoodnetwork@katuma-production:~/apps/openfoodnetwork/current$ tail -f log/delayed_job.log
2021-04-20T10:04:21+0200: Placing Order for Proxy Order 4301
2021-04-20T10:04:24+0200: Placing Order for Proxy Order 4302
2021-04-20T10:04:28+0200: Placing Order for Proxy Order 4303
2021-04-20T10:04:42+0200: Issue in Subscription Order 206178: changes
2021-04-20T10:04:42+0200: Placing Order for Proxy Order 4304
2021-04-20T10:04:47+0200: Placing Order for Proxy Order 4305
2021-04-20T10:04:50+0200: Placing Order for Proxy Order 4306
2021-04-20T10:04:53+0200: Placing Order for Proxy Order 4307
2021-04-20T10:04:56+0200: Placing Order for Proxy Order 4309
2021-04-20T10:04:59+0200: Placing Order for Proxy Order 4310
2021-04-20T10:05:26+0200: Issue in Subscription Order 206189: changes
2021-04-20T10:05:26+0200: Placing Order for Proxy Order 4311
2021-04-20T10:05:42+0200: Issue in Subscription Order 206190: changes
2021-04-20T10:05:43+0200: Placing Order for Proxy Order 4312
2021-04-20T10:05:46+0200: Placing Order for Proxy Order 4313
2021-04-20T10:05:51+0200: Issue in Subscription Order 206192: changes
2021-04-20T10:05:51+0200: Placing Order for Proxy Order 4315
2021-04-20T10:06:11+0200: Issue in Subscription Order 206193: changes
2021-04-20T10:06:11+0200: Placing Order for Proxy Order 4316
2021-04-20T10:06:35+0200: Issue in Subscription Order 206194: changes
2021-04-20T10:06:35+0200: Placing Order for Proxy Order 4317
2021-04-20T10:07:04+0200: Issue in Subscription Order 206195: changes
2021-04-20T10:07:05+0200: Placing Order for Proxy Order 4318
2021-04-20T10:07:08+0200: Placing Order for Proxy Order 4319
2021-04-20T10:07:13+0200: Issue in Subscription Order 206197: changes
2021-04-20T10:07:13+0200: Placing Order for Proxy Order 4320
2021-04-20T10:07:19+0200: Placing Order for Proxy Order 4321
^Ap2021-04-20T10:07:49+0200: Issue in Subscription Order 206199: changes
2021-04-20T10:07:50+0200: Placing Order for Proxy Order 4322
2021-04-20T10:07:55+0200: Placing Order for Proxy Order 4323
2021-04-20T10:07:58+0200: Placing Order for Proxy Order 4324
2021-04-20T10:08:04+0200: Issue in Subscription Order 206202: changes
2021-04-20T10:08:05+0200: Placing Order for Proxy Order 4326
2021-04-20T10:08:09+0200: Placing Order for Proxy Order 4327
2021-04-20T10:08:23+0200: Issue in Subscription Order 206204: changes
2021-04-20T10:08:24+0200: Placing Order for Proxy Order 4328
2021-04-20T10:08:47+0200: Issue in Subscription Order 206205: changes
2021-04-20T10:08:47+0200: Placing Order for Proxy Order 4329
2021-04-20T10:08:55+0200: Issue in Subscription Order 206206: changes
2021-04-20T10:08:56+0200: Placing Order for Proxy Order 4330
2021-04-20T10:09:12+0200: Issue in Subscription Order 206208: changes
2021-04-20T10:09:12+0200: Placing Order for Proxy Order 4331
2021-04-20T10:09:16+0200: Placing Order for Proxy Order 4333
2021-04-20T10:09:20+0200: Placing Order for Proxy Order 4334
2021-04-20T10:09:28+0200: Issue in Subscription Order 206211: changes
2021-04-20T10:09:28+0200: Placing Order for Proxy Order 4335
2021-04-20T10:09:32+0200: Placing Order for Proxy Order 4336
2021-04-20T10:09:39+0200: Issue in Subscription Order 206213: changes
2021-04-20T10:09:40+0200: Placing Order for Proxy Order 4337
2021-04-20T10:09:44+0200: Placing Order for Proxy Order 4339
2021-04-20T10:09:48+0200: Placing Order for Proxy Order 4340
I, [2021-04-20T10:09:51.725467 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionPlacementJob (id=997351) COMPLETED after 586.5830
I, [2021-04-20T10:09:51.729872 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job HeartbeatJob (id=997350) RUNNING
I, [2021-04-20T10:09:51.741481 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job HeartbeatJob (id=997350) COMPLETED after 0.0115
I, [2021-04-20T10:09:51.745294 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionConfirmJob (id=997352) RUNNING
I, [2021-04-20T10:09:51.758496 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionConfirmJob (id=997352) COMPLETED after 0.0131
I, [2021-04-20T10:09:51.761557 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job HeartbeatJob (id=997353) RUNNING
I, [2021-04-20T10:09:51.769482 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job HeartbeatJob (id=997353) COMPLETED after 0.0078
I, [2021-04-20T10:09:51.773197 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionConfirmJob (id=997354) RUNNING
I, [2021-04-20T10:09:51.787310 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionConfirmJob (id=997354) COMPLETED after 0.0140
I, [2021-04-20T10:09:51.790678 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionPlacementJob (id=997355) RUNNING
I, [2021-04-20T10:09:51.800020 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job SubscriptionPlacementJob (id=997355) COMPLETED after 0.0092
I, [2021-04-20T10:09:51.803590 #28054]  INFO -- : 2021-04-20T10:09:51+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper (id=997356) (queue=default) RUNNING
I, [2021-04-20T10:09:53.061849 #28054]  INFO -- : 2021-04-20T10:09:53+0200: [Worker(delayed_job host:katuma-production pid:28054)] Job ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper (id=997356) (queue=default) COMPLETED after 1.2580
I, [2021-04-20T10:09:53.065494 #28054]  INFO -- : 2021-04-20T10:09:53+0200: [Worker(delayed_job host:katuma-production pid:28054)] 7 jobs processed at 0.0119 j/s, 0 failed
</details>

We still need to assess:

1. If this issue only happens in Katuma. I wonder if bigger instances like the UK have much wider downtime windows or if it's a data-related thing that only happens there.
2. Is this a one-off event or the consequence of the increased adoption of subscriptions, that may lead us to think this is getting worse? can other correlations be established?

## Expected Behavior

The system should be available at all times regardless of subscriptions.

## Actual Behaviour

A 9min downtime window occurred during office hours, which could affect other critical background jobs such as user invitations.

## Steps to Reproduce

This could be reproduced by going through the same subscription and calling the Subs. Placement job in a staging server if needed.

## Workaround

There's none.

## Your Environment

Running release v3.6.6

## Possible Fix

Listing them from the more to least cost-effective

* Add a 2nd worker so we have High Availability. This may require extra RAM capacity to host it. See https://github.com/collectiveidea/delayed_job/tree/v4.1.9#running-jobs to create a worker pool.
* Split `SubscriptionPlacementJob` into more granular work units so we can segment its processing and allow other jobs to be executed in between.

**Comment by sauloperez** [comment]
Interesting, I checked UK's log, and the first and last log message for the 2021-04-21 are:

```
2021-04-21T05:00:05+0000: Placing Order for Proxy Order 15257
...
2021-04-21T05:06:06+0000: Placing Order for Proxy Order 22544
```

with a total of 119 log lines => 119 placed orders and none failed.

**Comment by Matt-Yorkley** [comment]
We did have some issues around those subscriptions jobs the other day, notes here: https://github.com/openfoodfoundation/openfoodnetwork/issues/7449#issuecomment-824038623

TLDR; we process tons of jobs all at the same time, which can be a bit crazy.

**Comment by sauloperez** [comment]
Katuma's stats definitely give room for another DJ worker

```sh
$ free -m
              total        used        free      shared  buff/cache   available
Mem:           7786        1242        1095         140        5448        6111
Swap:          2047           2        2045
```

**Comment by sauloperez** [comment]
> TLDR; when we process subscriptions we throw tons of jobs into the queue all at the same time, which can be a bit crazy.

In this case, looking at the log it was `SubscriptionPlacementJob` holding the worker for too long. If it was due to a burst in jobs, splitting in separate queues with different priority should also mitigate the issue.

**Comment by Matt-Yorkley** [comment]
Yeah, sounds good. A second worker and separate job queues is long overdue!

**Comment by sauloperez** [comment]
This needs to be revisited once https://github.com/openfoodfoundation/ofn-install/pull/744 gets merged and provisioned. The servers should all of a sudden have more memory available hopefully fitting another worker. The should also be a multiplier effect on those instances that run more app instances (Unicorn workers + DJ workers).

**Comment by Matt-Yorkley** [comment]
I think we can close this since we switched to Sidekiq? We have 5 workers now effectively, and the whole thing is a lot quicker and more performant.
